libs/sdk-ui-dashboard/src/model/queryServices/queryWidgetFilters.ts

```diff
@@ -0,0 +1,418 @@
+
+function refMatchesMdObject(ref: ObjRef, mdObject: IMetadataObject, type?: ObjectType): boolean {
+    return (
+        areObjRefsEqual(ref, mdObject.ref) ||
+        areObjRefsEqual(ref, idRef(mdObject.id, type)) ||
+        areObjRefsEqual(ref, uriRef(mdObject.uri))
+    );
+}
+
+async function loadDisplayFormsMetadata(
+    ctx: DashboardContext,
+    refs: ObjRef[],
+): Promise<IAttributeDisplayFormMetadataObject[]> {
+    if (!refs.length) {
+        return [];
+    }
+
+    return ctx.backend.workspace(ctx.workspace).attributes().getAttributeDisplayForms(refs);
+}
+
+function* getOrLoadDisplayFormsMetadata(
+    ctx: DashboardContext,
+    refs: ObjRef[],
+): SagaIterator<IAttributeDisplayFormMetadataObject[]> {
+    // first try getting as much as possible from catalog, there is a good chance the data is already there
+    const catalogDisplayForms: ReturnType<typeof selectCatalogAttributeDisplayForms> = yield select(
+        selectCatalogAttributeDisplayForms,
+    );
+
+    // for any ref not in catalog, load it from server (probably something from non-production dataset)
+    const refsMissing = refs.filter(
+        (ref) =>
+            !catalogDisplayForms.some((displayForm) => refMatchesMdObject(ref, displayForm, "displayForm")),
+    );
+
+    const dataForMissingRefs: PromiseFnReturnType<typeof loadDisplayFormsMetadata> = yield call(
+        loadDisplayFormsMetadata,
+        ctx,
+        refsMissing,
+    );
+
+    const allDisplayForms = [...catalogDisplayForms, ...dataForMissingRefs];
+
+    return refs.map((ref): IAttributeDisplayFormMetadataObject => {
+        const match = allDisplayForms.find((displayForm) =>
+            refMatchesMdObject(ref, displayForm, "displayForm"),
+        );
+        // if this bombs the logic above is broken
+        invariant(match);
+
+        return match;
+    });
+}
+
+interface IFilterDisplayFormPair {
+    filter: Exclude<IFilter, IDateFilter>;
+    displayForm: IAttributeDisplayFormMetadataObject | undefined;
+}
+
+interface IFilterDateDatasetPair {
+    filter: IDateFilter;
+    dateDataset: ICatalogDateDataset | undefined;
+}
+
+function* loadDisplayFormsForNonDateFilters(
+    ctx: DashboardContext,
+    filters: Exclude<IFilter, IDateFilter>[],
+): SagaIterator<IFilterDisplayFormPair[]> {
+    const refs = filters.map(filterObjRef);
+
+    const mdObjects = yield call(getOrLoadDisplayFormsMetadata, ctx, compact(refs));
+
+    let mdObjectPointer = 0;
+    let filterPointer = 0;
+
+    const result: IFilterDisplayFormPair[] = [];
+
+    while (filterPointer < filters.length) {
+        const filter = filters[filterPointer];
+        const hasObjRef = !!filterObjRef(filter);
+
+        if (hasObjRef) {
+            result.push({
+                displayForm: mdObjects[mdObjectPointer],
+                filter,
+            });
+            mdObjectPointer++;
+        } else {
+            result.push({
+                displayForm: undefined,
```

Comment:
Is this valid? (when displayForm can be undefined?)

Reply:
You are right it cannot happen, but this was handling also non-attribute filters (MFV and Ranking) and they in fact do not have display forms. I changed the logic so that this admittedly confusing part is gone.